### PR TITLE
Upgrade module cqrs command

### LINKED ISSUE #36977: Admin API - Update module

### PR DESCRIPTION
**Pull request by dylanDenizonPresta**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | Update module CQRS command
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | CI green
| UI Tests          | ~
| Fixed issue or discussion?     | Fixes #36977
| Related PRs       | ~
| Sponsor company   | ~

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop` or `8.2.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

**Comment by ps-jarvis** [comment]
Hello @dylanDenizonPresta!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

File: src/Adapter/Module/CommandHandler/UpdateModuleHandler.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+use PrestaShop\PrestaShop\Core\Domain\Module\Command\UpdateModuleCommand;
+use PrestaShop\PrestaShop\Core\Domain\Module\CommandHandler\UpdateModuleHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Module\Exception\ModuleAlreadyUpToDateException;
+use PrestaShop\PrestaShop\Core\Domain\Module\Exception\ModuleNotInstalledException;
+use PrestaShop\PrestaShop\Core\Domain\Module\Exception\CannotUpdateModuleException;
+use PrestaShop\PrestaShop\Core\Module\ModuleManager;
+use PrestaShop\PrestaShop\Core\Module\ModuleRepository;
+
+#[AsCommandHandler]
+class UpdateModuleHandler implements UpdateModuleHandlerInterface
```
Comment (jolelievre) [inline src/Adapter/Module/CommandHandler/UpdateModuleHandler.php]:
```suggestion
class UpgradeModuleHandler implements UpdateModuleHandlerInterface
```

File: src/Adapter/Module/CommandHandler/UpdateModuleHandler.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+{
+    public function __construct(
+        protected ModuleManager $moduleManager,
+        protected ModuleRepository $moduleRepository,
+    ) {
+    }
+
+    public function handle(UpdateModuleCommand $command): void
```
Comment (jolelievre) [inline src/Adapter/Module/CommandHandler/UpdateModuleHandler.php]:
```suggestion
    public function handle(UpgradeModuleCommand $command): void
```

File: tests/Integration/Behaviour/Features/Context/Domain/ModuleFeatureContext.php [hunk 3]
```diff
@@ -221,6 +231,20 @@ public function uploadModule(string $sourceType, string $sourceGiven, TableNode
         Module::resetStaticCache();
     }
 
+    /**
+     * @When I update module :technicalName
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/ModuleFeatureContext.php]:
```suggestion
     * @When I upgrade module :technicalName
```

File: tests/Integration/Behaviour/Features/Context/Domain/ModuleFeatureContext.php [hunk 4]
```diff
@@ -221,6 +231,20 @@ public function uploadModule(string $sourceType, string $sourceGiven, TableNode
...
+     */
+    public function updateModule(string $technicalName): void
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/ModuleFeatureContext.php]:
```suggestion
    public function upgradeModule(string $technicalName): void
```

File: src/Adapter/Module/CommandHandler/UpdateModuleHandler.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        if(!$module->canBeUpgraded()) {
+            throw new ModuleAlreadyUpToDateException('Module is already up to date.');
+        }
+
+        $result = $this->moduleManager->upgrade($technical_name);
+
+        if (!$result) {
+            throw new CannotUpdateModuleException('Technical error occurred while updating module.');
```
Comment (jolelievre) [inline src/Adapter/Module/CommandHandler/UpdateModuleHandler.php]:
```suggestion
            throw new CannotUpgradeModuleException('Technical error occurred while updating module.');
```

File: src/Core/Domain/Module/CommandHandler/UpdateModuleHandlerInterface.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Module\CommandHandler;
+
+use PrestaShop\PrestaShop\Core\Domain\Module\Command\UpdateModuleCommand;
+
+interface UpdateModuleHandlerInterface
```
Comment (jolelievre) [inline src/Core/Domain/Module/CommandHandler/UpdateModuleHandlerInterface.php]:
```suggestion
interface UpgradeModuleHandlerInterface
```

File: tests/Integration/Behaviour/Features/Scenario/Module/update_module.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Module/update_module.feature:1]: Files to be renamed into `upgrade_module.feature`

File: tests/Integration/Behaviour/Features/Scenario/Module/update_module.feature [hunk 7]
```diff
@@ -0,0 +1,69 @@
+# ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s module --tags update-module
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Module/update_module.feature]:
```suggestion
# ./vendor/bin/behat -c tests/Integration/Behaviour/behat.yml -s module --tags upgrade-module
```

File: tests/Integration/Behaviour/Features/Scenario/Module/update_module.feature [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+@reset-test-modules-before-feature
+@restore-all-tables-before-feature
+@clear-cache-before-feature
+@reset-test-modules-after-feature
+@update-module
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Module/update_module.feature]:
```suggestion
@upgrade-module
```

**Review by jolelievre — COMMENTED** [review]
Thanks @dylanDenizonPresta 

All good to me, now it's only missing the inclusion of your ps_apiresources branch for the new endpoints and we'll be done

**Review by jolelievre — DISMISSED** [review]
Thanks @dylanDenizonPresta

**Comment by jolelievre** [comment]
I commented a modification required on the module https://github.com/PrestaShop/ps_apiresources/pull/36#pullrequestreview-2446563261 You can also run the CS fixer on the module to validate its CI

Once it's done and you updated your branch on the module you can update it on this core branches by running this command `composer update prestashop/ps_apiresources` this will update the commit used in `composer.lock` that needs to be committed here as well

**Review by jolelievre — APPROVED** [review]
Thanks @dylanDenizonPresta

**Review by boherm — APPROVED** [review]

**Pull request merged**